velox/exec/tests/ArrowStreamTest.cpp

```diff
@@ -0,0 +1,240 @@
+          markGettingSchemaFailed_(markGettingSchemaFailed) {}
+
+    int getNext(struct ArrowArray* outArray) {
+      if (vectorIndex_ < vectors_.size()) {
+        exportToArrow(vectors_[vectorIndex_], *outArray, pool_.get());
+        vectorIndex_ += 1;
+        outArray->release = bridgeRelease;
+      } else {
+        // End of stream. Mark the array released.
+        outArray->release = nullptr;
+      }
+      return markGettingNextFailed_ ? 1 : 0;
+    }
+
+    int getArrowSchema(ArrowSchema& out) {
+      exportToArrow(BaseVector::create(type_, 0, pool_.get()), out);
+      out.release = bridgeSchemaRelease;
+      return markGettingSchemaFailed_ ? 2 : 0;
+    }
+
+   private:
+    std::shared_ptr<memory::MemoryPool> pool_;
+    std::vector<RowVectorPtr> vectors_;
+    TypePtr type_;
+    int vectorIndex_ = 0;
+    bool markGettingNextFailed_;
+    bool markGettingSchemaFailed_;
+  };
+
+  /// For the export from Arrow reader to Arrow array stream.
+  class ExportedArrayStream {
+   public:
+    struct PrivateData {
+      explicit PrivateData(std::shared_ptr<ArrowReader> reader)
+          : reader_(std::move(reader)) {}
+
+      std::shared_ptr<ArrowReader> reader_;
+      std::string lastError_;
+
+      PrivateData() = default;
+    };
+
+    explicit ExportedArrayStream(struct ArrowArrayStream* stream)
+        : stream_(stream) {}
+
+    int getSchema(struct ArrowSchema* outSchema) {
+      return reader()->getArrowSchema(*outSchema);
+    }
+
+    int getNext(struct ArrowArray* outArray) {
+      return reader()->getNext(outArray);
+    }
+
+    const char* getLastError() {
+      const auto& lastError = privateData()->lastError_;
+      return lastError.empty() ? nullptr : lastError.c_str();
+    }
+
+    void release() {
+      if (stream_->release == nullptr) {
+        // Array stream is released.
+        return;
+      }
+      VELOX_CHECK_NOT_NULL(privateData());
+      delete privateData();
+      // Mark stream as released.
+      stream_->release = nullptr;
+    }
+
+    static int staticGetSchema(
+        struct ArrowArrayStream* stream,
+        struct ArrowSchema* outSchema) {
+      ExportedArrayStream self{stream};
+      return self.toError(self.getSchema(outSchema));
+    }
+
+    static int staticGetNext(
+        struct ArrowArrayStream* stream,
+        struct ArrowArray* outArray) {
+      ExportedArrayStream self{stream};
+      return self.toError(self.getNext(outArray));
+    }
+
+    static void staticRelease(struct ArrowArrayStream* stream) {
+      ExportedArrayStream{stream}.release();
+    }
+
+    static const char* staticGetLastError(struct ArrowArrayStream* stream) {
+      return ExportedArrayStream{stream}.getLastError();
+    }
+
+   private:
+    PrivateData* privateData() {
+      return reinterpret_cast<PrivateData*>(stream_->private_data);
+    }
+    const std::shared_ptr<ArrowReader>& reader() {
+      return privateData()->reader_;
+    }
+
+    int toError(int code) {
```

Comment:
It might be cleaner to make this method static and have it return std::string for the error message.

```
std::string toErrorMessage(ErrorCode errorCode);


... 
privateData()->lastError_ = toErrorMessage(errorCode);
```

Reply:
Fixed.